### PR TITLE
MGDSTRM-8525: add topic config policy enhancement end to end tests

### PR DESCRIPTION
**Pull request by k-wall**
****Do not merge until FleetShard 0.27 is released****

Tests ensuring that the create and alter topic have the topic policy enforced correctly.

### CODE REVIEW

**Comment by showuon** [comment]
Thanks @k-wall , I'll take a look next Monday.

File: src/test/java/io/managed/services/test/kafka/KafkaInstanceAPITest.java [hunk 1]
```diff
@@ -165,6 +172,145 @@ public void testFailToCreateTopicIfItAlreadyExist() {
...
+
+                {true, newCE().key("segment.bytes").value(Integer.toString(fiftyMi))},
+                {true, newCE().key("segment.bytes").value(Integer.toString(fiftyMi + 1))},
+                {false, newCE().key("segment.bytes").value(Integer.toString(fiftyMi - 1))},
+                {false, newCE().key("segment.bytes").value(Integer.toString(1))},
+
+                {true, newCE().key("segment.index.bytes").value(Integer.toString(tenMi))},
+                {false, newCE().key("segment.index.bytes").value("1")},
+
+                //{false, newCE().key("segment.jitter.ms").value("0")},  // TODO - the specification says "We should prevent users from setting it", but the code does enforce that.
```
Comment (k-wall) [inline src/test/java/io/managed/services/test/kafka/KafkaInstanceAPITest.java:241]: @showuon the shows the implementation of the policy disagrees with the epic brief.

Comment (showuon) [reply src/test/java/io/managed/services/test/kafka/KafkaInstanceAPITest.java:241]: Oops, it's my bad. I misread the JIRA description. Yes, it should be `immutable`. I'll update it. Thanks for the test and catch it.

File: src/test/java/io/managed/services/test/kafka/KafkaInstanceAPITest.java [hunk 2]
```diff
@@ -165,6 +172,145 @@ public void testFailToCreateTopicIfItAlreadyExist() {
...
+
+                {false, newCE().key("message.format.version").value("3.0")},
+                {false, newCE().key("message.format.version").value("2.8")},
+                {false, newCE().key("message.format.version").value("2.1")},
+
+                {true, newCE().key("min.cleanable.dirty.ratio").value("0.5")},
+                {false, newCE().key("min.cleanable.dirty.ratio").value("0")},
+                {false, newCE().key("min.cleanable.dirty.ratio").value("1")},
+
+                //{true, newCE().key("min.insync.replicas").value(desiredBrokerCount > 2 ? "2" : "1")}, // TODO the specification says we should permit 2 for standard and 1 for developer
```
[the remark this answers is not in the document]
Comment (showuon) [reply src/test/java/io/managed/services/test/kafka/KafkaInstanceAPITest.java:230]: Yes, I noticed that, but has a question waiting to be answered. In our policy, we said it is default to 2 or 1 (standard or develop). But why can't user set `min.insync.replicas` to 3 in standard case?

And then, I found we already had this policy implemented in managed topic policy[1]. The validation logic is like this:
```
broker configured `min.insync.replicas` <= provided `min.insync.replicas` <= broker `default.replication.factor`
```
ex: broker min.insync.replicas=2, broker default.replication.factor=3. So, user can set min.insync.replicas to 2 or 3

I think that makes more sense. But still want to raise this question in our team channel. Another question is, the `default.replication.factor` is only allowed to set to 3, even in 6 brokers. Is that reasonable? IMO, we should only set lower bound (3) to ensure replicas are spreading across all AZs. If user wants create 6 replicas, that should be allowable. Waiting for the response in group room.

[1] https://github.com/bf2fc6cc711aee1a0c2a/kas-broker-plugins/blob/main/src/main/java/io/bf2/kafka/topic/ManagedKafkaCreateTopicPolicy.java#L71-L95

**Review by grdryn — APPROVED** [review]
LGTM, thanks @k-wall.

File: src/test/java/io/managed/services/test/kafka/KafkaInstanceAPITest.java [hunk 3]
```diff
@@ -165,6 +171,142 @@ public void testFailToCreateTopicIfItAlreadyExist() {
...
+        int messageSizeLimit, desiredBrokerCount;
+        try {
+            messageSizeLimit = KafkaMgmtApiUtils.getMessageSizeLimit(kafkaMgmtApi, kafka);
+            desiredBrokerCount = KafkaMgmtApiUtils.getDesiredBrokerCount(kafkaMgmtApi, kafka);
+        } catch (Exception e) {
+            // Fallback for kas-installer installed environments, see: https://github.com/bf2fc6cc711aee1a0c2a/kas-installer/issues/202
+            LOGGER.warn("Failed to read metrics, falling back to constants instead");
+            messageSizeLimit = 1048588;
+            desiredBrokerCount = 3;
+        }
```
Comment (showuon) [inline src/test/java/io/managed/services/test/kafka/KafkaInstanceAPITest.java:192]: nice!

**Review by showuon — COMMENTED** [review]
Nice tests. Left some comments.

**Pull request merged**